### PR TITLE
Task 1: code fixes and review

### PR DESCRIPTION
**Pull request by aceTmobile**
- Identified problems in code
- suggested improvements to code & why explanations
- noted accessibility issues & solutions
- fixed all tests

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
```
Comment (catfireparty) [inline .eslintrc:1]: Is this file a result of your upgrading Angular?

I notice you haven't also upgraded Nx. Do you know how to do this?

(Upgrading Angular without also upgrading Nx can be a problem)

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (catfireparty) [inline .vscode/launch.json]: Did you mean to commit this?

File: CODE_REVIEW.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+Readability
+- books.effects.spec (the current test description "loadbooks$ | it should work" could be a bit more descriptive by stating a specific expectation such as "it should return a list of books if a valid query is provided")
+- repeat for other specs where "it should work" & other vague descriptions are used
+
+DRY Principle
+- since API Endpoints are used in multiple places I would make them CONSTANTS instead of hard coded strings which can be proned to errors -> benefits track, change & debug in a single location
+- repeat for all areas where hard coded strings are used mutiple times
+
+
+Acccessibility
```
Comment (catfireparty) [inline CODE_REVIEW.md:25]: What about the lighthouse results? It doesn't read as though you've followed the instructions here.

File: CODE_REVIEW.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+- Books Reducer › valid Books actions › failedAddToReadingList should undo book addition to the state
+- Books Reducer › valid Books actions › failedRemoveFromReadingList should undo book removal from the state
+
+Readability
+- books.effects.spec (the current test description "loadbooks$ | it should work" could be a bit more descriptive by stating a specific expectation such as "it should return a list of books if a valid query is provided")
+- repeat for other specs where "it should work" & other vague descriptions are used
+
+DRY Principle
+- since API Endpoints are used in multiple places I would make them CONSTANTS instead of hard coded strings which can be proned to errors -> benefits track, change & debug in a single location
+- repeat for all areas where hard coded strings are used mutiple times
```
Comment (catfireparty) [inline CODE_REVIEW.md:14]: You've missed quite a lot here. These general points could apply to many code bases.

File: apps/okreads/browser/src/app/app.component.html [hunk 5]
```diff
@@ -5,6 +5,7 @@
         <h1>okreads</h1>
         <div>
           <button
+            aria-label="reading list"
```
Comment (catfireparty) [inline apps/okreads/browser/src/app/app.component.html]: Are you sure this button actually needs an `aria-label`?

(Hint: it doesn't. Why not?)

Comment (catfireparty) [reply apps/okreads/browser/src/app/app.component.html]: What about the button on line 27 below?

File: apps/okreads/browser/src/app/app.component.scss [hunk 6]
```diff
@@ -14,7 +14,7 @@ h1 {
   position: sticky;
   top: 0;
   z-index: 100;
-  background-color: $pink-accent;
+  background-color: $pink-dark;
```
Comment (catfireparty) [inline apps/okreads/browser/src/app/app.component.scss:17]: What precipitated this change?

File: libs/books/data-access/src/lib/+state/books.effects.ts [hunk 7]
```diff
@@ -5,14 +5,15 @@ import { catchError, map, switchMap } from 'rxjs/operators';
...
+import { API_ENDPOINT } from '@tmo/books/data-access';
 
 @Injectable()
 export class BooksEffects {
   searchBooks$ = createEffect(() =>
     this.actions$.pipe(
       ofType(BooksActions.searchBooks),
       switchMap((action) =>
-        this.http.get<Book[]>(`/api/books/search?q=${action.term}`).pipe(
+        this.http.get<Book[]>(API_ENDPOINT.GET_BOOKS_BY_QUERY+action.term).pipe(
```
Comment (catfireparty) [inline libs/books/data-access/src/lib/+state/books.effects.ts]: While this works, I personally don't find it as readable as the previous version. It's also less flexible. How would you implement this if there were multiple query params?

File: libs/books/data-access/src/lib/+state/reading-list.effects.ts [hunk 8]
```diff
@@ -5,14 +5,15 @@ import { of } from 'rxjs';
...
+import { API_ENDPOINT } from '@tmo/books/data-access';
 
 @Injectable()
 export class ReadingListEffects implements OnInitEffects {
   loadReadingList$ = createEffect(() =>
     this.actions$.pipe(
       ofType(ReadingListActions.init),
       exhaustMap(() =>
-        this.http.get<ReadingListItem[]>('/api/reading-list').pipe(
+        this.http.get<ReadingListItem[]>(API_ENDPOINT.READING_LIST).pipe(
```
Comment (catfireparty) [inline libs/books/data-access/src/lib/+state/reading-list.effects.ts]: What might be better than putting the endpoint into a constant?

Could you put the entire call into another service, perhaps?

File: libs/books/data-access/src/lib/+state/reading-list.reducer.ts [hunk 9]
```diff
@@ -47,11 +47,17 @@ const readingListReducer = createReducer(
       error: action.error
     };
   }),
-  on(ReadingListActions.addToReadingList, (state, action) =>
-    readingListAdapter.addOne({ bookId: action.book.id, ...action.book }, state)
+  on(ReadingListActions.addToReadingList, (state, action) => {
+    return readingListAdapter.addOne({ bookId: action.book.id, ...action.book }, state)}
```
Comment (catfireparty) [inline libs/books/data-access/src/lib/+state/reading-list.reducer.ts]: Don't change the convention half way through the file, there's no need for these bracketed return statements.

File: libs/books/feature/src/lib/book-search/book-search.component.html [hunk 10]
```diff
@@ -7,7 +7,7 @@
       placeholder="Search for books to add to your reading list"
       formControlName="term"
     />
-    <button mat-icon-button matSuffix>
+    <button mat-icon-button matSuffix aria-label="search icon">
```
Comment (catfireparty) [inline libs/books/feature/src/lib/book-search/book-search.component.html]: If this an appropriate label? Why would a partially sighted user need to know that this is an icon?

_Is_ it an icon?

File: libs/books/feature/src/lib/book-search/book-search.component.html [hunk 11]
```diff
@@ -47,7 +47,7 @@
 </ng-container>
 <ng-template #empty>
   <div class="empty">
-    <p>
+    <p id="searchTip">
```
Comment (catfireparty) [inline libs/books/feature/src/lib/book-search/book-search.component.html]: This doesn't seem to be referenced anywhere else. Generally if you're going to use ids, don't use camelCase, use kebab-case. If this is for testing, you want to make it `data-testid="..."`

Comment (catfireparty) [reply libs/books/feature/src/lib/book-search/book-search.component.html]: Hint: is the usage of an anchor tag on line 51 appropriate? What should it be instead?

File: libs/books/feature/src/lib/total-count/total-count.component.ts [hunk 12]
```diff
@@ -1,16 +1,29 @@
...
 
   constructor(private readonly store: Store) {}
 
   ngOnInit(): void {}
+  
+  ngAfterViewChecked(): void {
+    this.configureAccessibility()
+  }
+
+  configureAccessibility() {
```
Comment (catfireparty) [inline libs/books/feature/src/lib/total-count/total-count.component.ts]: You don't need to do this kind of DOM manipulation at all.

You can bind to the aria label quite simply as `<my-tag [attr.aria-label]="myvalue">` or simply `<my-tag aria-label="...">` if you're assigning a literal.

File: libs/shared/utils-accessibility/.eslintrc [hunk 13]
```diff
@@ -0,0 +1 @@
+{ "extends": "../../../.eslintrc", "rules": {}, "ignorePatterns": ["!**/*"] }
```
Comment (catfireparty) [inline libs/shared/utils-accessibility/.eslintrc]: This entire library is probably pointless.

File: libs/shared/utils-accessibility/src/lib/shared-utils-accessibility.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  id
+}
+
+export interface SelectorQuery {
+  selector: string;
+  type: SelectorType;
+  ariaLabelValue: string;
+}
+
+export function setAriaLabel(selectorQuery: SelectorQuery) {
```
Comment (catfireparty) [inline libs/shared/utils-accessibility/src/lib/shared-utils-accessibility.ts]: Angular has ways of doing all of this kind of manipulation.

File: tslint.json [hunk 15]
```diff
@@ -57,7 +57,7 @@
       true,
       {
         "enforceBuildableLibDependency": true,
-        "allow": [],
+        "allow": ["@tmo/shared"],
```
Comment (catfireparty) [inline tslint.json]: What's this enabling?

You might want to read up on tags -> https://nx.dev/latest/angular/structure/monorepo-tags

**Review by catfireparty — COMMENTED** [review]
You've missed quite a lot in the code. Take a closer look at the component classes and see how they could be (should be) improved. Pay attention to pipes, observables and lifecycle methods.

File: .eslintrc (hunk 1 above)
Comment (aceTmobile) [reply .eslintrc:1]: upgraded Nx

File: CODE_REVIEW.md (hunk 4 above)
Comment (aceTmobile) [reply CODE_REVIEW.md:14]: updated please review new changes. thanks

File: CODE_REVIEW.md (hunk 3 above)
Comment (aceTmobile) [reply CODE_REVIEW.md:25]: lighthouse results added

File: apps/okreads/browser/src/app/app.component.scss (hunk 6 above)
Comment (aceTmobile) [reply apps/okreads/browser/src/app/app.component.scss:17]: accessibility contrast flagged in lighthouse report. (Reading List button)

File: apps/okreads/browser/src/app/app.component.html (hunk 5 above)
Comment (aceTmobile) [reply apps/okreads/browser/src/app/app.component.html]: Removed: will override the visible text.

File: .vscode/launch.json (hunk 2 above)
Comment (aceTmobile) [reply .vscode/launch.json]: reverted

File: libs/books/feature/src/lib/book-search/book-search.component.html (hunk 11 above)
Comment (aceTmobile) [reply libs/books/feature/src/lib/book-search/book-search.component.html]: reverted

File: libs/shared/utils-accessibility/.eslintrc (hunk 13 above)
Comment (aceTmobile) [reply libs/shared/utils-accessibility/.eslintrc]: removed

File: tslint.json (hunk 15 above)
Comment (aceTmobile) [reply tslint.json]: reverted

File: libs/books/data-access/src/lib/+state/books.effects.ts (hunk 7 above)
Comment (aceTmobile) [reply libs/books/data-access/src/lib/+state/books.effects.ts]: reverted

File: libs/books/data-access/src/lib/+state/reading-list.reducer.ts (hunk 9 above)
Comment (aceTmobile) [reply libs/books/data-access/src/lib/+state/reading-list.reducer.ts]: reverted

File: libs/books/feature/src/lib/book-search/book-search.component.html (hunk 10 above)
Comment (aceTmobile) [reply libs/books/feature/src/lib/book-search/book-search.component.html]: updated - removed icon text

**Comment by aceTmobile** [comment]
@jdpearce Hi, changes per mr feedback are complete can you re-review? thanks.

**Comment by jaysoo** [comment]
LGTM

**Pull request merged**